### DISCUSSION TITLE
PrviderNotFoundException

### DISCUSSION THREAD

**Discussion by dadzio008** [Pi4J/pi4j#126] [Q&A]
I have problem creating analog and digital input. Every time i see error 
Exception in thread "Thread-0" com.pi4j.provider.exception.ProviderNotFoundException: Pi4J provider [pigpio-digital-input] could not be found.  Please include this 'provider' JAR in the classpath.
	at com.pi4j.provider.impl.DefaultRuntimeProviders.get(DefaultRuntimeProviders.java:238)
	at com.pi4j.provider.impl.DefaultProviders.get(DefaultProviders.java:147)
	at com.pi4j.context.Context.provider(Context.java:240)
	at If.ifs(If.java:19)
	at Main.run(Main.java:12)
What's going on? Anyone can help?

**Comment by vidikn** [comment]
I have the same issue, did you manage to fix it?

**Comment by eitch** [comment]
Hi guys. We have seen issues when using fat jars, i.e. all the projects dependencies are extracted in to the projects own jar. If this is the case with you, then i am currently sad to say that that won't work. The pi4j-v2 uses a automatic lookup service for the providers and after the creation of a fat jar, the service files defining these providers is reduced to one, as they have the same name.

Can you please modify your project to copy the dependencies into a sub folder, e.g. lib/ and then reference those in the class/module path?

**Reply by dadzio008** [reply]
but i don't use module-info. when i tried that and creating subfolder for dependencies there were a few other strange errors.

**Reply by dadzio008** [reply]
i'm beginner in java and i dont know what to do

**Reply by wolff77** [reply]
well, I'm suffering from the same problem so I've put out that jar by making a `<scope>provided</scope>` to the maven file.
I've also added the jars to a sub folder and loaded them using the classpath when executing java, but the error just stays.. or do I miss a jar ? I've put only pi4j-core-2.1.0.jar and pi4j-library-linuxfs-2.1.0.jar into the lib subfolder ... and to be honest, I try to use the linuxfs for i2c instead of gpio

**Reply by wolff77** [reply]
Ah, Just forget about my last comment, I've seen that I've included the wrong dependency .... only the library from maven repo instead of the plugin. My error, now it's working.
However @dadzio008 try to put that scope thing into your dependency; pull the pi4j jars out of the .m2 directory in your development computer and add this to a folder you like; after that call java with the -cp parameter and include there your main jar and the folder with a *. that should do the trick

**Reply by peterivarsson** [reply]
I have the same problem.
com.pi4j.provider.exception.ProviderNotFoundException: Pi4J provider [linuxfs-i2c] could not be found.

Could you explain more in detail how you fixed this issue.

**Reply by peterivarsson** [reply]
What classpath:s do you use

**Comment by nujabness** [comment]
Hello there, I’m trying to read a moisture sensor the purpuse is to have a percentage measure.

I’ve read a lot of topic on PI4J but I still don’t understand how I can do this.

I read earlier: "The PiGPIO plugin is a set of I/O providers for just about every supported I/O type with the exception of analog inputs and outputs."

Does this mean that my sensor can only be used with the digital provider ?

I tried many things with AnalogInputConfigBuilder but I always get: ProviderNotFoundException: Pi4J provider IO type [ANALOG_INPUT] could not be found.

Also, I noticed at the execution that there is no analog provider: 
PROVIDERS: [12] "I/O Providers" <com.pi4j.provider.impl.DefaultProviders>
├─ANALOG_INPUT: [0] <com.pi4j.io.gpio.analog.AnalogInputProvider>
├─ANALOG_OUTPUT: [0] <com.pi4j.io.gpio.analog.AnalogOutputProvider>

is there any solution to get an analogical value from a moisture sensor with PI4J ? Or did someone manage to make this example work : 

https://github.com/Pi4J/pi4j-examples/blob/master/src/main/java/com/pi4j/example/gpio/analog/input/AnalogInputExample.java


Edit: 
My bad, I found that :    
        // The 'RaspberryPi' does not inherently support analog I/O
        // hardware, but with an attached ADC (Analog to Digital Converter)
        // or DAC (Digital to Analog converter) chip attached to a
        // data bus (I2C/SPI) you may wish to use Pi4J to read/write
        // to these analog hardware interfaces.

**Reply by TobiasReich** [reply]
Would be great to know. The existence of this class indicates for me there is a way for analog input.
Did you find anything in this regard?
